lib/elixir/test/elixir/kernel/guard_test.exs

```diff
@@ -0,0 +1,361 @@
+Code.require_file "../test_helper.exs", __DIR__
+
+defmodule Kernel.GuardTest do
+  use ExUnit.Case, async: true
```

Comment:
You can't use `capture_io(:stderr)` with `async: true`. My recommendation would be to move those warnings to "test/kernel/warnings_test.exs".

Reply:
Are the examples there the preferred way to write tests that test compiling code? ie:

```elixir
  test "unused variable in defguard" do
    capture_err(fn ->
      Code.eval_string """
      defmodule Sample do
        defguard foo(bar, baz) when bar
      end
      """
    end) =~ "variable \"baz\" is unused"
  after
    purge Sample
  end
```

Reply:
Those are testing warnings. For testing errors, you can probably use assert_raise:

```elixir
assert_raise error, message, fn ->
  defmodule Foo do
    defguard ...
  end
end
```